### PR TITLE
Feature signing user as teacher, and finding all teachers

### PR DESCRIPTION
**Pull request by MateuszBednarczyk**
- created teacher controller
- added new role
- added signasteacherdto
- added teacherdto
- added userisalreadyateacherexception exception class
- added find all roles by roletype method
- created teacher service class and created find all teachers method and add teacher role to user method
- created teacher service implementation and implemented methods
- added hashCode and equals methods to role entity

### CODE REVIEW

File: src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+@RequestMapping("api/v1/teachers")
+@RequiredArgsConstructor
+public class TeacherController {
+
+    private final TeacherService teacherService;
+
+    @GetMapping("/all")
+    public ResponseEntity<List<TeacherDTO>> findAllTeachers() {
+        return new ResponseEntity<>(teacherService.findAllTeachers(), HttpStatus.OK);
+    }
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java:23]: I would recommend implementing pagination on this endpoint. We don't know how many users will use our app and retrieving, for example, 10000 users would be very time consuming.
Also, I would leave the endpoint's address at simply `api/v1/teachers` instead of `api/v1/teachers/all`, but that is just my opinion and it doesn't really matter that much.

File: src/main/java/pl/simpleascoding/tutoringplatform/domain/user/Role.java [hunk 2]
```diff
@@ -22,4 +23,17 @@ public class Role {
...
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Role role = (Role) o;
+        return roleType == role.roleType;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(roleType);
+    }
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/domain/user/Role.java:38]: I think it's better to use Lombok's `@EqualsAndHashCode`.

File: src/main/java/pl/simpleascoding/tutoringplatform/dto/TeacherDTO.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package pl.simpleascoding.tutoringplatform.dto;
+
+import org.hibernate.annotations.Immutable;
+
+@Immutable
+public record TeacherDTO(String name, String surname, String email) {
+}
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/dto/TeacherDTO.java:7]: Do we need to have a separate class for teacher info? Wouldn't it be better to use our existing `UserDTO` class instead?

File: src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.CONFLICT)
+public class UserIsAlreadyATeacherException extends IllegalArgumentException {
+    public UserIsAlreadyATeacherException() {
+        super("This is already a teacher");
+    }
+}
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java:11]: I would recommend including the user's username in the message, as well as extracting the message into a final string field.
Example: `private static final String MESSAGE_USER_ALREADY_A_TEACHER = "User %s is already a teacher";`

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            throw new UserIsAlreadyATeacherException();
+        }
+    }
+
+    private void completeTeacherDtoListWithDtoObjects(List<TeacherDTO> teacherDTOList) {
+        for (Role teacherRole : roleRepository.findRoleByRoleType(RoleType.TEACHER)) {
+            User teacher = teacherRole.getUser();
+            teacherDTOList.add(new TeacherDTO(teacher.getName(), teacher.getSurname(), teacher.getEmail()));
+        }
+    }
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java:57]: This method looks very strange to me, why is it populating a provided list instead of returning one?
Also, try to use MapStruct where possible, please refer to these materials on how to do it:
[MapStruct webpage](https://mapstruct.org/)
[ReviewModelMapper class from our project](https://github.com/Simple-as-Coding/tutoring-platform/blob/main/src/main/java/pl/simpleascoding/tutoringplatform/service/review/ReviewModelMapper.java)
[ReviewServiceImpl class from our project](https://github.com/Simple-as-Coding/tutoring-platform/blob/d21d32e98d9fe5ade638b49fb23660e381b432b6/src/main/java/pl/simpleascoding/tutoringplatform/service/review/ReviewServiceImpl.java#L51)

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    private Role createTeacherRoleEntity() {
+        return new Role(RoleType.TEACHER);
+    }
+}
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java:62]: Is this method necessary?

File: src/main/java/pl/simpleascoding/tutoringplatform/repository/RoleRepository.java [hunk 7]
```diff
@@ -2,7 +2,10 @@
 
 import org.springframework.data.jpa.repository.JpaRepository;
 import pl.simpleascoding.tutoringplatform.domain.user.Role;
+import pl.simpleascoding.tutoringplatform.domain.user.RoleType;
+
+import java.util.List;
 
 public interface RoleRepository extends JpaRepository<Role, Long> {
-    //todo
+    List<Role> findRoleByRoleType(RoleType roleType);
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/repository/RoleRepository.java:10]: Please rename this method to `findRolesByRoleType`, as it's searching for multiple Role objects

Comment (Eukon05) [reply src/main/java/pl/simpleascoding/tutoringplatform/repository/RoleRepository.java:10]: We should rework our roles system to allow searching for users directly in `UserRepository`, but for now it's a valid solution.

File: src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 package pl.simpleascoding.tutoringplatform.domain.user;
 
 public enum RoleType {
-    ADMIN, USER
+    ADMIN, USER, TEACHER
```
Comment (Eukon05) [inline src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java:4]: We should rename the `USER` role to `STUDENT` instead, but since it's not related to this feature, it will/should be handled in a separate pull request

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java (hunk 6 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java:62]: In my opinion - yes, because one method, should do only one operation, or return information about object.

File: src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java (hunk 8 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java:4]: You're right, but I thought that it should be done in commits, that refers to student

File: src/main/java/pl/simpleascoding/tutoringplatform/domain/user/Role.java (hunk 2 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/domain/user/Role.java:38]: As we talked at discord, it could be problematic in the future

File: src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java (hunk 1 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java:23]: I'll think about it. but to be honest, I don't think that we'll have 10000 users. If yes, we should use other framework to take care about databases. Hibernate is good and flexible but it's also made for small applications (like 2000/3000 users).

File: src/main/java/pl/simpleascoding/tutoringplatform/dto/TeacherDTO.java (hunk 3 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/dto/TeacherDTO.java:7]: You're right. I thought that we just don't need teacher id, but it may be important to frontend in sending requests

File: src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java (hunk 4 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java:11]: you're right

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java (hunk 5 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java:57]: right

File: src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java (hunk 1 above)
Comment (MateuszBednarczyk) [reply src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java:23]: About endpoint - 'all' refers to all teachers, and only all teachers.

Comment (dawciobiel) [reply src/main/java/pl/simpleascoding/tutoringplatform/api/TeacherController.java:23]: @Eukon, I do not agree. We will talk on the communicator on this issue.

File: src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java (hunk 8 above)
Comment (dawciobiel) [reply src/main/java/pl/simpleascoding/tutoringplatform/domain/user/RoleType.java:4]: Agree. We will modify it with "student feature".

File: src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java (hunk 4 above)
Comment (dawciobiel) [reply src/main/java/pl/simpleascoding/tutoringplatform/exception/UserIsAlreadyATeacherException.java:11]: Agree, cuz we were doing like that in other cases.

**Review by dawciobiel — CHANGES_REQUESTED** [review]
Changes are required we will speak about it on voice chat.

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    private final RoleRepository roleRepository;
+    private final UserService userService;
+    private final UserModelMapper userModelMapper;
+
+    @Override
+    @Transactional
+    public void addTeacherRoleToUser(SignAsTeacherDTO requestDTO) {
+        User user = userService.getUserByUsername(requestDTO.username());
+        Role roleTeacher = createTeacherRoleEntity();
+        isUserAlreadyTeacher(user, roleTeacher);
```
Comment (dawciobiel) [inline src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java]: Name of this method is wrong. Name of the method right now is the question - that mean it should return some value. But it doesn't.

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+@RequiredArgsConstructor
+public class TeacherServiceImpl implements TeacherService {
+
+    private final RoleRepository roleRepository;
+    private final UserService userService;
+    private final UserModelMapper userModelMapper;
+
+    @Override
+    @Transactional
+    public void addTeacherRoleToUser(SignAsTeacherDTO requestDTO) {
```
Comment (dawciobiel) [inline src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java:29]: method should return proper value  - some DTO with code/status/message etc..

File: src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Override
+    public List<UserDTO> findAllTeachers() {
+        List<UserDTO> teacherDTOList = new ArrayList<>();
+        completeTeacherDtoListWithDtoObjects(teacherDTOList);
+
+        return teacherDTOList;
+    }
+
+    private void isUserAlreadyTeacher(User user, Role roleTeacher) {
```
Comment (dawciobiel) [inline src/main/java/pl/simpleascoding/tutoringplatform/service/teacher/TeacherServiceImpl.java]: method should return proper value  - some DTO with code/status/message etc..

**Review by dawciobiel — APPROVED** [review]
Some of the code will be refactored - DTO object as return values from some method gonna be added.
